bootloader/grub.go

```diff
@@ -379,11 +384,80 @@ func (g *grub) ManagedAssets() []string {
 	}
 }
 
-// CommandLine returns the kernel command line composed of the built-in
-// list and extra arguments passed in arguments. The command line may be
-// different when using a bootloader in the recovery partition.
+// CommandLine returns the kernel command line composed of system run
+// mode arguments, built-in bootloader specific static arguments
+// followed by any extra arguments configured by snapd. The command line
+// may be different when using a bootloader in the recovery partition.
 //
 // Implements ManagedAssetsBootloader for the grub bootloader.
-func (g *grub) CommandLine(extra []string) (string, error) {
-	return "", fmt.Errorf("not implemented")
+func (g *grub) CommandLine(modeArgs, extraArgs string) (string, error) {
+	// we do not trust the on disk asset, use the built-in one
+	assetName := "grub.cfg"
+	isRecovery := false
+	if g.recovery {
+		assetName = "grub-recovery.cfg"
+		isRecovery = true
+	}
+	currentBootConfig := filepath.Join(g.dir(), "grub.cfg")
+	edition, err := editionFromDiskConfigAsset(currentBootConfig)
+	if err != nil {
+		return "", fmt.Errorf("cannot obtain edition number of current boot config: %v", err)
+	}
+	staticCmdline := staticCommandLineForGrubAssetEdition(assetName, edition)
+	// split the mode arguments and make sure they are sane for given
+	// run/recovery mode
+	mode, err := modeArgsForGrub(modeArgs, isRecovery)
+	if err != nil {
+		return "", err
+	}
+	args, err := strutil.KernelCommandLineSplit(staticCmdline + " " + extraArgs)
+	if err != nil {
+		return "", fmt.Errorf("cannot use badly formatted kernel command line: %v", err)
+	}
+	// join all argument with a single space, see
+	// grub-core/lib/cmdline.c:grub_create_loader_cmdline() for reference,
+	// arguments are separated by a single space, the space after last is
+	// replaced with terminating NULL
+	return strings.Join(append(mode, args...), " "), nil
+}
+
+// staticCommandLineForGrubAssetEdition fetches a static command line for given
+// grub asset edition
+func staticCommandLineForGrubAssetEdition(asset string, edition uint) string {
+	cmdline := assets.SnippetForEdition(fmt.Sprintf("%s:static-cmdline", asset), edition)
+	if cmdline == nil {
+		return ""
+	}
+	return string(cmdline)
+}
+
+// modeArgsForGrub validates the snapd mode arguments for the kernel command
+// line and splits them into individual arguments.
+func modeArgsForGrub(modeArgs string, recoveryAsset bool) ([]string, error) {
```

Comment:
this feels like something that boot and not each bootloader should care about. Maybe CommandLine should take separate modeArg and optional systemArg explicitly ?

Reply:
Yes, that would mean we can skip the sorting business entirely.